src/tools/tidy/src/rustdoc_html_ids.rs

```diff
@@ -0,0 +1,164 @@
+        Some(nb) => *nb += 1,
+        None => {
+            eprintln!(
+                "=> ID `{}` in file `{}` at line {} is missing from `init_id_map`",
+                id,
+                path.display(),
+                line_nb + 1,
+            );
+            *bad = true;
+        }
+    }
+}
+
+fn check_ids(
+    path: &Path,
+    f: &str,
+    ids: &mut HashMap<String, usize>,
+    regex: &Regex,
+    bad: &mut bool,
+    small_section_header_checked: &mut usize,
+) {
+    let mut is_checking_small_section_header = None;
+
+    for (line_nb, line) in f.lines().enumerate() {
+        let trimmed = line.trim_start();
+        // We're not interested in comments or doc comments.
+        if trimmed.starts_with("//") {
+            continue;
+        } else if let Some(start_line) = is_checking_small_section_header {
+            if line_nb == start_line + 2 {
+                check_id(path, trimmed.split('"').skip(1).next().unwrap(), ids, line_nb, bad);
+                is_checking_small_section_header = None;
+            }
+        } else if trimmed.starts_with("write_small_section_header(") {
+            // This function is used to create section: the second argument of the function is an
+            // ID and we need to check it as well, hence this specific check...
+            if trimmed.contains(',') {
+                // This is a call made on one line, so we can simply check it!
+                check_id(path, trimmed.split('"').skip(1).next().unwrap(), ids, line_nb, bad);
+            } else {
+                is_checking_small_section_header = Some(line_nb);
+            }
+            *small_section_header_checked += 1;
+            continue;
+        }
+        for cap in regex.captures_iter(line) {
+            check_id(path, &cap[1], ids, line_nb, bad);
+        }
+    }
+}
+
+pub fn check(path: &Path, bad: &mut bool) {
+    // matches ` id="blabla"`
+    let regex = Regex::new(r#"[\s"]id=\\?["']([^\s\\]+)\\?["'][\s\\>"{]"#).unwrap();
+
+    println!("Checking rustdoc IDs...");
+    let mut ids = extract_ids(&path.join(ID_MAP_PATH), bad);
+    let mut small_section_header_checked = 0;
+    if *bad {
+        return;
+    }
+    super::walk(
+        &path.join("librustdoc/html"),
+        &mut |path| super::filter_dirs(path),
+        &mut |entry, contents| {
+            let path = entry.path();
+            let file_name = entry.file_name();
+            if path.extension() == Some(OsStr::new("html"))
+                || (path.extension() == Some(OsStr::new("rs")) && file_name != "tests.rs")
+            {
+                check_ids(path, contents, &mut ids, &regex, bad, &mut small_section_header_checked);
+            }
+        },
+    );
+    if small_section_header_checked == 0 {
+        eprintln!(
+            "No call to the `write_small_section_header` function was found. Was it renamed?",
+        );
+        *bad = true;
+    }
+    for (id, nb) in ids {
+        if IDS_USED_IN_JS.iter().any(|i| i == &id) {
```

Comment:
nit: I think this would be simpler:

```suggestion
        if IDS_USED_IN_JS.contains(&id) {
```